tests/multi_threaded.rs

```diff
@@ -0,0 +1,146 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+mod common;
+
+use common::{CounterInnerMutSubscriber, CounterSubscriber, CounterSubscriberWithData};
+use event_manager::{EventManager, EventSubscriber, MutEventSubscriber, SubscriberOps};
+use std::sync::{Arc, Mutex};
+use std::thread;
+
+// Showcase how you can manage subscribers of different types using the same event manager
+// in a multithreaded context.
+#[test]
+fn test_diff_type_subscribers() {
+    let mut event_manager = EventManager::<Arc<Mutex<dyn MutEventSubscriber + Send>>>::new()
+        .expect("Cannot create event manager.");
+
+    // The `CounterSubscriberWithData` expects to receive a number as a parameter that represents
+    // the number it can use as its inner Events data.
+    let data_subscriber = Arc::new(Mutex::new(CounterSubscriberWithData::new(1)));
+    data_subscriber.lock().unwrap().trigger_all_counters();
+
+    let counter_subscriber = Arc::new(Mutex::new(CounterSubscriber::new()));
+    counter_subscriber.lock().unwrap().trigger_event();
+
+    // Saving the ids allows to modify the subscribers in the event manager context (i.e. removing
+    // the subscribers from the event manager loop).
+    let ds_id = event_manager.add_subscriber(data_subscriber);
+    let cs_id = event_manager.add_subscriber(counter_subscriber);
+
+    let thread_handle = thread::spawn(move || {
+        // We cannot run the event manager loop for ever. Let's just run it for 100 iterations.
+        for _ in 0..100 {
+            // When the event manager loop exits, it will return the number of events it
+            // handled.
+            let ev_count = event_manager.run().unwrap();
+            // We are expecting the ev_count to be:
+            // 4 = 3 (events triggered from data_subscriber) + 1 (event from counter_subscriber).
+            assert_eq!(ev_count, 4);
+        }
+
+        // One really important thing is to always remove the subscribers once you don't need them
+        // any more.
```

Comment:
Maybe it's worth mentioning this is especially important when the `EventManger` keeps running, because if it also gets dropped, then subscribers are dropped as well anyway.

Reply:
I'll make it more explicit.

Reply:
Another thing to keep in mind is that remove still needs to be called even when the `EventManager` does not run anymore. Should we actually implement Drop on EventManger instead?

Reply:
It doesn't look like we need to have an explicit `Drop` implementation for `EventManager` (also please see the other comment related to `Drop`).

Reply:
The last comment is super confusing for me. Why wouldn't we need explicit drop for eventmanager? Who else calls remove_subscriber if not the event_manager who has ownership of the subscribers? `remove_subscriber` could also be called by the user of the library, but at the scope of this code, event manager should manage its own cleanup, right?